interlok-core/src/main/java/com/adaptris/core/services/splitter/PooledSplitJoinService.java

```diff
@@ -0,0 +1,437 @@
+import static com.adaptris.core.util.ServiceUtil.discardNulls;
+import java.util.Iterator;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import com.adaptris.annotation.AdapterComponent;
+import com.adaptris.annotation.AdvancedConfig;
+import com.adaptris.annotation.ComponentProfile;
+import com.adaptris.annotation.DisplayOrder;
+import com.adaptris.annotation.InputFieldDefault;
+import com.adaptris.core.AdaptrisMessage;
+import com.adaptris.core.CoreException;
+import com.adaptris.core.EventHandler;
+import com.adaptris.core.EventHandlerAware;
+import com.adaptris.core.Service;
+import com.adaptris.core.ServiceException;
+import com.adaptris.core.ServiceImp;
+import com.adaptris.core.ServiceWrapper;
+import com.adaptris.core.services.aggregator.MessageAggregator;
+import com.adaptris.core.services.splitter.ServiceWorkerPool.Worker;
+import com.adaptris.core.util.Args;
+import com.adaptris.core.util.ExceptionHelper;
+import com.adaptris.core.util.LifecycleHelper;
+import com.adaptris.core.util.ManagedThreadFactory;
+import com.adaptris.interlok.util.CloseableIterable;
+import com.adaptris.interlok.util.Closer;
+import com.adaptris.util.NumberUtils;
+import com.adaptris.util.TimeInterval;
+import com.thoughtworks.xstream.annotations.XStreamAlias;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import lombok.Setter;
+
+/**
+ * Implementation of the Splitter and Aggregator enterprise integration pattern.
+ *
+ * <p>
+ * This supersedes {@link SplitJoinService} and {@link PoolingSplitJoinService} as our preferred
+ * service for handling split/aggregation since it has more predictable peformance characteristics
+ * in constrained environments where the numbers of messages that are generated and aggregated can
+ * be large/unknown.
+ * </p>
+ * <p>
+ * This service splits a message according to the configured {@link MessageSplitter} implementation,
+ * executes the configured {@link Service} and subsequently aggregates all the messages back using
+ * the configured {@link MessageAggregator} implementation
+ * <p>
+ * <p>
+ * A pool of {@link com.adaptris.core.Service} instances is maintained and re-used for each message;
+ * the cost of initialisation for the wrapped service, is incurred during this service's
+ * initialisation phase.
+ * </p>
+ * <p>
+ * Aggregation may start happening as soon as messages are available to be aggregated (i.e. a split
+ * message has been operated on) using the
+ * {@link MessageAggregator#aggregate(AdaptrisMessage, Iterable)} method. Performance characterstics
```

Comment:
Javadoc typo: It should be "Performance characteristics" instead of "Performance characterstics"